crates/optimism/payload/src/tree_validator.rs

```diff
@@ -0,0 +1,203 @@
+    StateCommitmentProvider, StateProviderFactory, StateReader,
+};
+use reth_trie::HashedPostState;
+use std::sync::Arc;
+
+/// Common trait bounds for the provider type used throughout [`OpPayloadValidator`]
+pub trait OpProvider:
+    DatabaseProviderFactory<Provider: BlockReader + BlockNumReader + HeaderProvider>
+    + BlockReader
+    + BlockNumReader
+    + StateProviderFactory
+    + StateReader
+    + StateCommitmentProvider
+    + HashedPostStateProvider
+    + HeaderProvider<Header = <OpPrimitives as reth_primitives_traits::NodePrimitives>::BlockHeader>
+    + Clone
+    + Unpin
+    + 'static
+{
+}
+
+/// Automatic implementation for types that satisfy the bounds
+impl<P> OpProvider for P where
+    P: DatabaseProviderFactory<Provider: BlockReader + BlockNumReader + HeaderProvider>
+        + BlockReader
+        + BlockNumReader
+        + StateProviderFactory
+        + StateReader
+        + StateCommitmentProvider
+        + HashedPostStateProvider
+        + HeaderProvider<
+            Header = <OpPrimitives as reth_primitives_traits::NodePrimitives>::BlockHeader,
+        > + Clone
+        + Unpin
+        + 'static
+{
+}
+
+/// Optimism-specific payload validator that uses [`TreePayloadValidator`] for common validation
+/// logic.
+#[derive(Debug)]
+pub struct OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+{
+    /// Reusable validation logic provider
+    tree_validator: TreePayloadValidator<OpPrimitives, P, C>,
+    /// Chain spec for hardfork checks
+    chain_spec: Arc<Spec>,
+}
+
+impl<P, C, Spec> OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+    Spec: OpHardforks,
+{
+    /// Creates a new Optimism payload validator.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        provider: P,
+        consensus: Arc<dyn FullConsensus<OpPrimitives, Error = ConsensusError>>,
+        evm_config: C,
+        config: TreeConfig,
+        payload_processor: PayloadProcessor<OpPrimitives, C>,
+        precompile_cache_map: PrecompileCacheMap<SpecFor<C>>,
+        invalid_block_hook: Box<dyn InvalidBlockHook<OpPrimitives>>,
+        chain_spec: Arc<Spec>,
+    ) -> Self {
+        let tree_validator = TreePayloadValidator::new(
+            provider,
+            consensus,
+            evm_config,
+            config,
+            payload_processor,
+            precompile_cache_map,
+            100, // invalid_headers_cache_size
+            invalid_block_hook,
+            Default::default(), // metrics
+        );
+
+        Self { tree_validator, chain_spec }
+    }
+}
+
+impl<P, C, Spec> PayloadValidator for OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+    Spec: OpHardforks + EthereumHardforks + Send + Sync + 'static,
+{
+    type Block = OpBlock;
```

Comment:
this is more restrictive than

https://github.com/paradigmxyz/reth/blob/2c5a967898aeb6018d3f3a9a67a206b192289b7a/crates/optimism/node/src/engine.rs#L121-L121

but this should also ago away if we restructure this